src/flamenco/runtime/program/fd_bpf_loader_v3_program.c

```diff
@@ -190,1352 +330,1149 @@ fd_bpf_loader_v3_user_execute( fd_exec_instr_ctx_t ctx ) {
+      fd_native_cpi_create_account_meta( payer_key,       1U, 1U, &acct_metas[ 0UL ] ); 
+      fd_native_cpi_create_account_meta( programdata_key, 1U, 1U, &acct_metas[ 1UL ] );
+      fd_native_cpi_create_account_meta( buffer_key,      0U, 1U, &acct_metas[ 2UL ] );
+      fd_pubkey_t signers[ 2UL ];
+      ulong signers_cnt = 2UL;
+      signers[ 0UL ] = *payer_key;
+      signers[ 1UL ] = *programdata_key;
+
+      err = fd_native_cpi_execute_system_program_instruction( instr_ctx, &instr, acct_metas, 
+                                                              3UL, signers, signers_cnt );
+      if( err ) {
         return err;
       }
     } FD_SCRATCH_SCOPE_END;
 
-    ulong total_size = PROGRAMDATA_METADATA_SIZE + instruction.inner.deploy_with_max_data_len.max_data_len;
-    fd_borrowed_account_t * programdata_rec = NULL;
-    int modify_err = fd_instr_borrowed_account_modify( &ctx, programdata_acc, total_size, &programdata_rec );
-    FD_TEST( modify_err == FD_ACC_MGR_SUCCESS );
-    fd_account_meta_t * meta = programdata_rec->meta;
-    uchar * acct_data        = programdata_rec->data;
-
-    fd_bpf_upgradeable_loader_state_t program_data_acc_loader_state = {
-      .discriminant = fd_bpf_upgradeable_loader_state_enum_program_data,
-      .inner.program_data.slot = ctx.slot_ctx->slot_bank.slot,
-      .inner.program_data.upgrade_authority_address = (fd_pubkey_t *)authority_acc
-    };
-
-    fd_bincode_encode_ctx_t encode_ctx;
-    encode_ctx.data = acct_data;
-    encode_ctx.dataend = acct_data + fd_bpf_upgradeable_loader_state_size(&program_data_acc_loader_state);
-    if ( fd_bpf_upgradeable_loader_state_encode( &program_data_acc_loader_state, &encode_ctx ) ) {
-      FD_LOG_ERR(("fd_bpf_upgradeable_loader_state_encode failed"));
-      fd_memset( acct_data, 0, fd_bpf_upgradeable_loader_state_size(&program_data_acc_loader_state) );
-      return FD_EXECUTOR_INSTR_ERR_GENERIC_ERR;
+    /* https://github.com/anza-xyz/agave/blob/574bae8fefc0ed256b55340b9d87b7689bcdf222/programs/bpf_loader/src/lib.rs#L644-L665 */
+    /* Load and verify the program bits */
+    if( BUFFER_METADATA_SIZE>buffer_data_len ) {
+      FD_LOG_WARNING(( "Buffer data length is too large" ));
+      return FD_EXECUTOR_INSTR_ERR_ACC_DATA_TOO_SMALL;
     }
 
-    meta->dlen = PROGRAMDATA_METADATA_SIZE + instruction.inner.deploy_with_max_data_len.max_data_len;
-    meta->info.executable = 0;
-    fd_memcpy(&meta->info.owner, fd_solana_bpf_loader_upgradeable_program_id.key, sizeof(fd_pubkey_t));
-    meta->info.lamports = fd_rent_exempt_minimum_balance(ctx.slot_ctx, meta->dlen);
-    meta->info.rent_epoch = 0;
-
-    buffer_data_len = fd_ulong_sat_sub(buffer_rec->meta->dlen, BUFFER_METADATA_SIZE);
+    if( buffer_data_offset>=buffer->meta->dlen ) {
+      FD_LOG_WARNING(( "Buffer data offset is out of bounds" ));
+      return FD_EXECUTOR_INSTR_ERR_ACC_DATA_TOO_SMALL;
+    }
 
```

Comment:
Where are these checks in Agave? These don't line up with the source - do you mind adding a permalink 🙏